packages/nimble-components/src/menu/index.ts

```diff
@@ -14,7 +27,387 @@ declare global {
+        return (
+            this.parentElement !== null
+            && isHTMLElement(this.parentElement)
+            && this.parentElement.getAttribute('role') === 'menuitem'
+        );
+    };
+
+    /**
+     * Focuses the first item in the menu.
+     *
+     * @public
+     */
+    public override focus(): void {
+        this.setFocus(0, 1);
+    }
+
+    /**
+     * Collapses any expanded menu items.
+     *
+     * @public
+     */
+    public collapseExpandedItem(): void {
+        if (this.expandedItem !== null) {
+            this.expandedItem.expanded = false;
+            this.expandedItem = null;
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public handleMenuKeyDown(e: KeyboardEvent): boolean {
+        if (e.defaultPrevented || this.menuItems === undefined) {
+            return false;
+        }
+        switch (e.key) {
+            case keyArrowDown:
+                // go forward one index
+                this.setFocus(this.focusIndex + 1, 1);
+                return false;
+            case keyArrowUp:
+                // go back one index
+                this.setFocus(this.focusIndex - 1, -1);
+                return false;
+            case keyEnd:
+                // set focus on last item
+                this.setFocus(this.menuItems.length - 1, -1);
+                return false;
+            case keyHome:
+                // set focus on first item
+                this.setFocus(0, 1);
+                return false;
+
+            default:
+                // if we are not handling the event, do not prevent default
+                return true;
+        }
+    }
+
+    /**
+     * if focus is moving out of the menu, reset to a stable initial state
+     * @internal
+     */
+    public handleFocusOut = (e: FocusEvent): void => {
+        if (
+            !this.contains(e.relatedTarget as Element)
+            && this.menuItems !== undefined
+        ) {
+            this.collapseExpandedItem();
+            // find our first focusable element
+            const focusIndex: number = this.menuItems.findIndex(
+                this.isFocusableElement
+            );
+            // set the current focus index's tabindex to -1
+            this.menuItems[this.focusIndex]?.setAttribute('tabindex', '-1');
+            // set the first focusable element tabindex to 0
+            this.menuItems[focusIndex]?.setAttribute('tabindex', '0');
+            // set the focus index
+            this.focusIndex = focusIndex;
+        }
+    };
+
+    private itemsChanged(
+        _oldValue: HTMLElement[],
+        newValue: HTMLElement[]
+    ): void {
+        // only update children after the component is connected and
+        // the setItems has run on connectedCallback
+        // (menuItems is undefined until then)
+        if (this.$fastController.isConnected && this.menuItems !== undefined) {
+            this.setItems();
+        }
+
+        this.mutationObserver.disconnect();
```

Comment:
I would probably be a good idea to disconnect the mutationObserver when the component is disconnected and start observing when the component is reconnected.

Reply:
I noticed that `setItems` is probably a better place to observe all the menu items, so I moved the calls there. Then I only had to `disconnect` in the `disconnectedCallback`.